File: rust/src/streaming/runtime/transfer/data_reader.rs
use std::iter::Map;
use std::collections::HashMap;
use zmq;
use cha

/// https://github.com/erickt/rust-zmq/blob/master/examples/msgsend/main.rs

struct DataReader {
    name: &str,
    zmq_context: &mut zmq::Context,
    channels: &Vec<(&str, &str, u32)>,
    sockets: HashMap<&str, &zmq::Socket>,
    cur_read_id: u32,
    running: bool
}

impl DataReader {

    fn new(name: &str, channels: Vec<(&str, &str, u32)>) -> DataReader {
        let mut ctx = zmq::Context::new();
        let mut sockets = HashMap::new();
        for channel_info in channels {
            let channel_id = channel_info.0;
            let source_ip = channel_info.1;
            let source_port = channel_info.2;
            let pull_socket = ctx.socket(zmq::PULL).unwrap();
            let addr = format!("tcp://{source_ip}:{source_port}");
            /// TODO pull_socket.setsockopt(zmq::LINGER, 0)
            pull_socket.connect(addr).unwrap();
            sockets.insert(channel_info[0], pull_socket)
        }

        DataReader {
            name: name,
            zmq_context: ctx,
            channels: &channels,
            sockets: &sockets,
            cur_read_id: 0,
            running: true,
        }
    }

    fn read_message(&self) -> ChannelMessage {
        let channel_id = &self.channels[&self.cur_read_id].0;
        let socket = &self.sockets[channel_id];
        let json_str: Option<&str> = None;

        while &self.running && json_str.is_none() {
            json_str = socket.recv_string(zmq::NOBLOCK)
        }

        let msg: ChannelMessage = serde_json::from_str(data)?;
        msg
    }

    fn close() {
        /// TODO
    }
}

File: volga/streaming/runtime/transfer/deprecated/data_reader.py
import logging
from typing import List, Optional

from volga.streaming.runtime.transfer.channel import Channel, ChannelMessage, RemoteChannel
from volga.streaming.runtime.transfer.deprecated.channel import Channel_DEP
from volga.streaming.runtime.transfer.deprecated.data_writer import TransportType

import zmq
import simplejson


logger = logging.getLogger("ray")


class DataReader_DEPR:
    def __init__(
        self,
        name: str,
        input_channels: List[Channel_DEP],
        transport_type: TransportType = TransportType.ZMQ_PUSH_PULL
    ):
        if transport_type not in [
            TransportType.ZMQ_PUSH_PULL,
        ]:
            raise RuntimeError(f'Unsupported transport {transport_type}')

        self.name = name
        self.input_channels = input_channels
        self.cur_read_id = 0
        self.running = True

        # TODO buffering
        # buffer pool impl https://github.com/Naman-Bhalla/dbms-buffer-pool-manager-python/tree/master/src
        self.sockets_and_contexts = {}
        for channel in self.input_channels:
            assert isinstance(channel, RemoteChannel)
            context = zmq.Context()
            # TODO set HWM
            socket = context.socket(zmq.PULL)
            socket.setsockopt(zmq.LINGER, 0)
            socket.connect(f'tcp://{channel.source_ip}:{channel.source_port}')
            self.sockets_and_contexts[channel.channel_id] = (socket, context)

    # TODO set timeout
    def read_message(self) -> Optional[ChannelMessage]:
        # TODO this should use a buffer?

        # round-robin read
        channel_id = None
        json_str = None
        while self.running and json_str is None:
            channel_id = self.input_channels[self.cur_read_id].channel_id
            socket = self.sockets_and_contexts[channel_id][0]
            try:
                json_str = socket.recv_string(zmq.NOBLOCK)
            except zmq.error.ContextTerminated:
                logger.info('zmq recv interrupt due to ContextTerminated')
                json_str = None
            except Exception as e:
                json_str = None
            self.cur_read_id = (self.cur_read_id + 1) % len(self.input_channels)

        # reader was stopped
        if json_str is None:
            assert self.running is False
            return None

        # TODO serialization perf
        msg = simplejson.loads(json_str)
        return msg

    def close(self):
        self.running = False
        # cleanup sockets and contexts for all channels
        for channel_id in self.sockets_and_contexts:
            socket = self.sockets_and_contexts[channel_id][0]
            context = self.sockets_and_contexts[channel_id][1]
            socket.close(linger=0)
            context.destroy(linger=0)
